File: backend/src/app/routers/audit.py
"""GET /api/audit — compliance-role-only, filterable, paginated audit trail
(E3-S3 AC1-AC5; api-contracts.md §13.1).

Only `GET` is ever registered on this router — no write verb exists anywhere
in this module, so FastAPI itself answers POST/PUT/PATCH/DELETE with its own
405 (AC3), never a hand-rolled one. `require_role("compliance")` is the sole
authorisation check (AC2, AC5); this router calls `domain.audit.service`
only, never `domain.audit.repository` directly (system-design.md D3).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from src.app.dependencies import CurrentUser, get_session, require_role
from src.domain.audit.service import list_audit_entries
from src.types.entities import AuditDetailValue

router = APIRouter(prefix="/api/audit", tags=["audit"])

DEFAULT_LIMIT = 50
MAX_LIMIT = 200


class AuditEntryResponse(BaseModel):
    """One `AuditLogEntry` row (api-contracts.md §13.1)."""

    id: int
    entity_type: str
    entity_id: str
    actor_id: int
    actor_role: str
    action: str
    timestamp: str
    details_json: dict[str, AuditDetailValue]


class AuditPageResponse(BaseModel):
    """The paginated envelope every `GET /api/audit` response uses."""

    total: int
    limit: int
    offset: int
    entries: list[AuditEntryResponse]


@router.get("", status_code=200)
def get_audit_entries(
    actor_id: int | None = Query(default=None),
    entity_type: str | None = Query(default=None),
    from_: str | None = Query(default=None, alias="from"),
    to: str | None = Query(default=None),
    limit: int = Query(default=DEFAULT_LIMIT, ge=1, le=MAX_LIMIT),
    offset: int = Query(default=0, ge=0),
    user: CurrentUser = Depends(require_role("compliance")),
    session: Session = Depends(get_session),
) -> AuditPageResponse:
    """Compliance-only paginated read of the audit trail (AC1, AC4)."""
    del user
    page = list_audit_entries(
        session,
        actor_id=actor_id,
        entity_type=entity_type,
        start=from_,
        end=to,
        limit=limit,
        offset=offset,
    )
    return AuditPageResponse(
        total=page.total,
        limit=page.limit,
        offset=page.offset,
        entries=[
            AuditEntryResponse(
                id=entry.id,
                entity_type=entry.entity_type.value,
                entity_id=entry.entity_id,
                actor_id=entry.actor_id,
                actor_role=entry.actor_role.value,
                action=entry.action.value,
                timestamp=entry.timestamp,
                details_json=entry.details_json,
            )
            for entry in page.entries
        ],
    )

File: frontend/src/api/audit.ts
/**
 * GET /api/audit — compliance-role-only, filterable, paginated audit trail
 * (E3-S4; api-contracts.md §13.1).
 */
import type { AuditAction, AuditDetailValue, AuditEntityType, Role } from '../types/entities';
import { apiRequest } from './client';

export interface AuditEntryResponse {
  id: number;
  entity_type: AuditEntityType;
  entity_id: string;
  actor_id: number;
  actor_role: Role;
  action: AuditAction;
  timestamp: string;
  details_json: Record<string, AuditDetailValue>;
}

export interface AuditPageResponse {
  total: number;
  limit: number;
  offset: number;
  entries: AuditEntryResponse[];
}

export interface AuditQuery {
  actor_id?: number;
  entity_type?: string;
  from?: string;
  to?: string;
  limit?: number;
  offset?: number;
}

/** GET /api/audit?... — omits any filter that is `undefined` or an empty
 * string, so an untouched filter never narrows the server-side query. */
export function getAuditEntries(query: AuditQuery = {}): Promise<AuditPageResponse> {
  const params = new URLSearchParams();
  if (query.actor_id !== undefined) {
    params.set('actor_id', String(query.actor_id));
  }
  if (query.entity_type !== undefined && query.entity_type !== '') {
    params.set('entity_type', query.entity_type);
  }
  if (query.from !== undefined && query.from !== '') {
    params.set('from', query.from);
  }
  if (query.to !== undefined && query.to !== '') {
    params.set('to', query.to);
  }
  if (query.limit !== undefined) {
    params.set('limit', String(query.limit));
  }
  if (query.offset !== undefined) {
    params.set('offset', String(query.offset));
  }
  const queryString = params.toString();
  const path = queryString.length > 0 ? `/api/audit?${queryString}` : '/api/audit';
  return apiRequest<AuditPageResponse>(path);
}
